### ISSUE TITLE
Babi-T2T decoder?

### ISSUE THREAD

**Issue by andreamad8** [MostafaDehghani/bAbI-T2T#4]
First of all, great work with the Uni Transformer. Then I have some question about the implementation and the hyperparameters:

- From the paper and the implementation, I miss how to use the transformer-decoder (if used). Indeed, I have just tried to use the transformer-encoder by concatenating [story, query] and taking the last vector in output from the encoder to predict over the vocabulary. In this way, Uni Transformer and Transformer worked pretty well (Uni Better as u mentioned in the paper). From the implementation, I don't get if you guys use also the decoder. If so how? could you give a bit more details

- I couldn't find the hyperparameter used in the experiments, are you going to release it somewhere? 

Thanks a lot for your great work, looking forward to hear from you. 

Best 

Andrea

**Comment by MostafaDehghani** [comment]
Hi @andreamad8,

For bAbI, similar to all the previous works, we treat the task as a classification task, where we take the vocabulary as the set of labels (which is somehow equivalent to one step decoding!). So the bAbI problem is kind of a T2T text2class problem.  You can find the code that I modified for decoding from bAbI transformer [here](https://github.com/MostafaDehghani/bAbI-T2T/blob/master/t2t_usr_dir/babi_transformer_decoder.py). 

Btw, we use the _top_ function from the _class label_ modality to go from encoded [story, query] to [answer]. It is, in fact, a projection from all the encoded inputs (not just the last one!) to the class label space. Check the code [here](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/layers/modalities.py#L812). 

About the hparams_set, you can use [universal_transformer_tiny](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/research/universal_transformer.py#L458) or [universal_transformer_tall](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/research/universal_transformer.py#L483) to get the results of the universal transformer that are reported in the paper.

**Comment by MostafaDehghani** [comment]
Hi @andreamad8,

For bAbI, similar to all the previous works, we treat the task as a classification task, where we take the vocabulary as the set of labels (which is somehow equivalent to one step decoding!). So the bAbI problem is kind of a T2T text2class problem. You can find the code that I modified for decoding from bAbI transformer here.

Btw, we use the top function from the class label modality to go from encoded [story, query] to [answer]. It is, in fact, a projection from all the encoded inputs (not just the last one!) to the class label space. Check the code [here](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/layers/modalities.py#L812).

About the hparams_set, you can use universal_transformer_tiny or universal_transformer_tall to get the results of the universal transformer that are reported in the paper.

**Comment by andreamad8** [comment]
Ok, I got the Hyper-Parameters and the top function. I still have some doubt about the decoding. 

From my understanding, I would just define a Transformer Encoder, and use the final matrix $H^t$ (at the last layer), using the top function (reduce-mean), to predict the class (i.e. Vocabulary). 

If a decoder is required, what would be the input token? SOS? PAD? 

I am a bit confused and sorry for asking many questions. I am just wondering if I fully get the flow. 

Thanks  

Andrea

**Comment by MostafaDehghani** [comment]
mmm...I actually didn't get what exactly you mean by "what would be the input token? SOS? PAD?".

The code for the Universal Transformer with the encoder **and the decoder**, that can be used for text2text problems (i.e. where you need decoding text from the model), is in the T2T repo. You can find the code that implemented the UT decoder [here](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/research/universal_transformer.py#L90). The point is that the Universal Transformer is pretty similar to the Transformer model in terms of the architecture and the use cases. So you can decode from UT similar to the way that you decode from Transformer.  Anyways, this is not the case for bAbI task as we have text2class setup and no decoder is used here.

Let me know if it wasn't the answer to your question!

**Comment by andreamad8** [comment]
Hi, 

oks, now I get. I was just confused about bAbI case (text2class), now everything clear.

I will close issue. 

Thanks again for your time

Andrea
